File: LC/Medium/875.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)
        ans = l
        
        while l <= r:
            m = (l+r) // 2
            time = 0
            for b in piles:
                time += ceil(b/m)
            if time <= h:
                r = m - 1
                ans = m
            else:
                l = m + 1
        
        return ans

File: LC/C++/Medium/875.cpp
#include <algorithm>
#include <cmath>
#include <iostream>

class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int left = 1;
        int right = *std::max_element(piles.begin(), piles.end());
        int ans = right;

        while (left <= right) {
            int guess = (left + right) / 2;
            double hours_taken = 0;
            for (auto& bananas : piles) {
                hours_taken += std::ceil((double)bananas / guess);
            }

            if (hours_taken <= h) {
                ans = min(ans, guess);
                right = guess - 1;
            } else {
                left = guess + 1;
            }
        }
        return ans;
    }
};
